### PR TITLE
Fix TensorFlow plugin operation without GPU

### PR DESCRIPTION
**Pull request by JanuszL**
- TensorFlow DALI plugin doesn't work without GPU as it forces
  synchronization after copying out the output data to the TF tensor
  what invokes CUDA call. This PR removes this synchronization when
  we copy to the non-pinned CPU buffer and when the DALI TF operator
  is placed on the CPU

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Category:

**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
- TensorFlow DALI plugin doesn't work without GPU as it forces
  synchronization after copying out the output data to the TF tensor
  what invokes CUDA call. This PR removes this synchronization when
  we copy to the non-pinned CPU buffer and when the DALI TF operator
  is placed on the CPU

### Affected modules and functionalities:
- DALI TF operator and dataset
- CPU only tests

### Key points relevant for the review:
- tests

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali_tf_plugin/dali_dataset_op.cc [hunk 1]
```diff
@@ -650,7 +650,10 @@ class DALIDatasetOp::Dataset::Iterator : public DatasetIterator<Dataset> {
 
       // Synchronize with the dataset()->stream_ when doing the last copy, so the outputs
       // are fully finished before we release the output buffers for reuse.
-      unsigned int wait_flag = (out_id == num_outputs - 1) ? DALI_ext_force_sync : DALI_ext_default;
+      // if the OP runs on the CPU the output memory is not pinned and we don't need to sync
+      unsigned int wait_flag = (i == dali_num_out - 1) ?
+                            (this->device_type_ == device_type_t::CPU ? 0 : DALI_ext_force_sync) :
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali_tf_plugin/dali_dataset_op.cc:655]: Or
```
unsigned int wait_flag =  this->device_type_ != device_type_t::CPU && (i == dali_num_out - 1) ? DALI_ext_force_sync : DALI_ext_default;
```

File: dali_tf_plugin/daliop.cc [hunk 2]
```diff
@@ -350,10 +350,12 @@ class DaliOp : public tf::OpKernel {
           break;
       }
 
-
       // Synchronize with the dataset()->stream_ when doing the last copy, so the outputs
       // are fully finished before we release the output buffers for reuse.
-      unsigned int wait_flag = (i == dali_num_out - 1) ? DALI_ext_force_sync : DALI_ext_default;
+      // if the OP runs on the CPU the output memory is not pinned and we don't need to sync
+      unsigned int wait_flag = (i == dali_num_out - 1) ?
```
Comment (jantonguirao) [inline dali_tf_plugin/daliop.cc:356]: same here

File: qa/TL0_cpu_only/test_pytorch.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# used pip packages
+pip_packages="nose numpy>=1.17 pillow torch numba scipy librosa==0.8.1 tensorflow-gpu"
+
+target_dir=./dali/test/python
+
+test_body() {
+  # CPU only test, remove CUDA from the search path just in case
+  export LD_LIBRARY_PATH=""
+  export PATH=${PATH/cuda/}
+  nosetests --verbose test_dali_cpu_only.py
```
Comment (jantonguirao) [inline qa/TL0_cpu_only/test_pytorch.sh:11]:
```suggestion
  nosetests --verbose -m '(?:^|[\b_\./-])[Tt]est.*pytorch*' test_dali_cpu_only.py
```
?

Comment (jantonguirao) [reply qa/TL0_cpu_only/test_pytorch.sh:11]: Or should we use `--attr 'pytorch'` vs. `--attr '!pytorch'` ? Right now we are running all tests under test_pytorch.sh

File: dali/test/python/test_dali_tf_plugin_cpu_only_dataset.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import nvidia.dali.types as types
+import nvidia.dali.plugin.tf as dali_tf
+from test_utils_tensorflow import skip_for_incompatible_tf
+import tensorflow as tf
+import random
+import numpy as np
+
+
+@pipeline_def()
+def get_datali_pipe(value):
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_tf_plugin_cpu_only_dataset.py]: Done

File: dali/test/python/test_dali_tf_plugin_cpu_only_dataset.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    data = types.Constant(value)
+    return data
+
+def get_data(batch_size, value):
+    pipe = get_datali_pipe(batch_size=batch_size, device_id=types.CPU_ONLY_DEVICE_ID, num_threads=1, value=value)
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_tf_plugin_cpu_only_dataset.py]: Done

File: dali/test/python/test_dali_tf_plugin_cpu_only_dataset.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+def test_dali_tf_dataset_cpu_only():
+    skip_for_incompatible_tf()
+    try:
+        tf.compat.v1.enable_eager_execution()
+    except:
+        pass
+
+    batch_size = 3
+    value = random.randint(0, 1000)
+    pipe = get_datali_pipe(batch_size=batch_size, device_id=types.CPU_ONLY_DEVICE_ID, num_threads=1, value=value)
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_dali_tf_plugin_cpu_only_dataset.py]: Done

File: dali_tf_plugin/dali_dataset_op.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali_tf_plugin/dali_dataset_op.cc]: Done

File: dali_tf_plugin/daliop.cc (hunk 2 above)
Comment (JanuszL) [reply dali_tf_plugin/daliop.cc]: Done

**Review by prak-nv — APPROVED** [review]

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**